scrapy/responsetypes.py

```diff
@@ -40,6 +43,9 @@ def __init__(self):
 
     def from_mimetype(self, mimetype):
         """Return the most appropriate Response class for the given mimetype"""
+        if isinstance(mimetype, bytes):
+            mimetype = mimetype.decode()
```

Comment:
Instead of doing this, use `mimetype.decode()` in the calling code, to keep the API of this method simple and unmodified (i.e. only allow strings as input).

Once you reimplement this method to rely on `xtractmime` methods instead of `CLASSES`, what you could do is support `bytes` as input but do the opposite: encode `mimetype` if it is `str`, and probably log a deprecation warning asking users to pass `bytes` instead.

I’m also thinking that it _might_ make sense to perform the changes we discussed for this method in a separate pull request, if you are OK with that. Separate, minimal pull requests (as opposed to single, bloated pull requests) are easier to review and merge, and I think this is a change that can be implemented separately.